src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs

```diff
@@ -1855,25 +1855,33 @@ private void LogConflict(Reference reference, string fusionName)
 
         #region StateFile
         /// <summary>
-        /// Reads the state file (if present) into the cache.
+        /// Reads the state file (if present) into the cache. If not present, attempts to read from CacheInputPaths, then creates a new cache if necessary.
         /// </summary>
-        private void ReadStateFile()
+        internal void ReadStateFile(GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo, Func<string, Guid> calculateMvid = null, Func<string, bool> fileExists = null)
         {
-            _cache = (SystemState)StateFileBase.DeserializeCache(_stateFile, Log, typeof(SystemState));
+            _cache = StateFileBase.DeserializeCache<SystemState>(_stateFile, Log);
 
-            // Construct the cache if necessary.
             if (_cache == null)
             {
-                _cache = new SystemState();
+                _cache = SystemState.DeserializePrecomputedCaches(AssemblyInformationCachePaths ?? Array.Empty<ITaskItem>(), Log, typeof(SystemState), getLastWriteTime, installedAssemblyTableInfo, calculateMvid, fileExists);
+            }
+            else
+            {
+                _cache.SetGetLastWriteTime(getLastWriteTime);
+                _cache.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
             }
         }
 
         /// <summary>
-        /// Write out the state file if a state name was supplied and the cache is dirty.
+        /// If CacheOutputPath is non-null, writes out a cache to that location. Otherwise, writes out the state file if a state name was supplied and the cache is dirty.
         /// </summary>
-        private void WriteStateFile()
+        internal void WriteStateFile(Func<string, Guid> calculateMvid = null)
         {
-            if (!string.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
+            if (!string.IsNullOrEmpty(AssemblyInformationCacheOutputPath))
+            {
+                _cache.SerializePrecomputedCache(AssemblyInformationCacheOutputPath, Log, calculateMvid);
+            }
+            else if (!string.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
```

Comment:
Does it make any sense to run the else branch even when `AssemblyInformationCacheOutputPath` is set?

Reply:
I don't think there's any need to. It wouldn't be a bad idea to switch the order of these; the first is expected to almost never be true, and if it is, you're creating a cache to send to users, and since that would likely only happen on a remote build machine, you probably don't care about local state.

Reply:
I guess my question would be "why are you creating a precomputed cache if you're just doing an incremental build"? The precomputed caches are for sending to customers, and they're only useful when building from an otherwise cache-free state. Note that if I remember correctly, serializing caches took ~7% of RAR's total time, and serializing two caches would double the time spent on it.